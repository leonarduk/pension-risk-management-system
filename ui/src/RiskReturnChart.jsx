import {
  Chart as ChartJS,
  LinearScale,
  PointElement,
  Tooltip,
  Legend,
} from 'chart.js';
import { Scatter } from 'react-chartjs-2';

ChartJS.register(LinearScale, PointElement, Tooltip, Legend);

export default function RiskReturnChart({ data }) {
  const chartData = {
    datasets: [
      {
        label: 'Risk vs Return',
        data: data.map((d) => ({ x: d.annual_std, y: d.annual_return, label: d.ticker })),
        backgroundColor: 'rgb(75, 192, 192)',
      },
    ],
  };

  const options = {
    plugins: {
      tooltip: {
        callbacks: {
          label: (ctx) => {
            const { label, x, y } = ctx.raw;
            return `${label}: σ=${(x * 100).toFixed(2)}%, μ=${(y * 100).toFixed(2)}%`;
          },
        },
      },
    },
    scales: {
      x: { title: { display: true, text: 'Annual Std Dev' } },
      y: { title: { display: true, text: 'Annual Return' } },
    },
  };

  return <Scatter data={chartData} options={options} />;
}
