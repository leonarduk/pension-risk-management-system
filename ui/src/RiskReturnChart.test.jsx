import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import { expect, test, vi, afterEach } from 'vitest';
import RiskReturnChart from './RiskReturnChart.jsx';

afterEach(() => {
  cleanup();
});

vi.mock('react-chartjs-2', () => ({
  Scatter: (props) => <pre data-testid="scatter-chart">{JSON.stringify(props)}</pre>,
}));

test('renders scatter chart with provided data', () => {
  const data = [{ ticker: 'AAPL', annual_std: 0.1, annual_return: 0.2 }];
  render(<RiskReturnChart data={data} />);
  const chart = screen.getByTestId('scatter-chart');
  expect(chart.textContent).toContain('"x":0.1');
  expect(chart.textContent).toContain('"y":0.2');
  expect(chart).toMatchSnapshot();
});

test('handles empty data', () => {
  render(<RiskReturnChart data={[]} />);
  expect(screen.getByTestId('scatter-chart')).toBeTruthy();
});
