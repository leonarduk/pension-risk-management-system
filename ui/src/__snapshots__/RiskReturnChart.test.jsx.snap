// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders scatter chart with provided data 1`] = `
<pre
  data-testid="scatter-chart"
>
  {"data":{"datasets":[{"label":"Risk vs Return","data":[{"x":0.1,"y":0.2,"label":"AAPL"}],"backgroundColor":"rgb(75, 192, 192)"}]},"options":{"plugins":{"tooltip":{"callbacks":{}}},"scales":{"x":{"title":{"display":true,"text":"Annual Std Dev"}},"y":{"title":{"display":true,"text":"Annual Return"}}}}}
</pre>
`;
